File: src/com/s16/drawing/PatternDrawable.java
package com.s16.drawing;

import java.util.HashMap;

import android.annotation.SuppressLint;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;

public class PatternDrawable extends Drawable {

	@SuppressLint("UseSparseArrays")
	private final HashMap<Integer, Bitmap> mBitmapCache = new HashMap<Integer, Bitmap>();
	
	private final Paint mDrawPaint;
	private final Paint mPaint;
	private int mBitmapWidth = 0;
	private int mBitmapHeight = 0;
	private Bitmap mSrcBitmap;
	private Bitmap mBitmap;
	private int numRectanglesHorizontal;
	private int numRectanglesVertical;
	
	public PatternDrawable(Bitmap bitmap) {
		mDrawPaint = new Paint();
		mPaint = new Paint();
		if (bitmap != null) {
			mSrcBitmap = bitmap;
			mBitmapWidth = mSrcBitmap.getWidth();
			mBitmapHeight = mSrcBitmap.getHeight();
		}
	}
	
	public PatternDrawable(Resources res, int id) {
		mDrawPaint = new Paint();
		mPaint = new Paint();
		if ((res != null) && (id > 0)) {
			mSrcBitmap = BitmapFactory.decodeResource(res, id);
			mBitmapWidth = mSrcBitmap.getWidth();
			mBitmapHeight = mSrcBitmap.getHeight();
		}
	}
	
	@Override
	public void draw(Canvas canvas) {
		canvas.drawBitmap(mBitmap, null, getBounds(), mDrawPaint);
	}

	@Override
	public void setAlpha(int alpha) {
		throw new UnsupportedOperationException("Alpha is not supported by this drawwable.");
	}

	@Override
	public void setColorFilter(ColorFilter cf) {
		throw new UnsupportedOperationException("ColorFilter is not supported by this drawwable.");
	}

	@Override
	public int getOpacity() {
		return 0;
	}
	
	@Override
	public int getIntrinsicWidth() {
        return mBitmapWidth;
    }
	
	@Override
	public int getIntrinsicHeight() {
        return mBitmapHeight;
    }
	
	@Override
	protected void onBoundsChange(Rect bounds) {
		super.onBoundsChange(bounds);
		
		int height = bounds.height();
		int width = bounds.width();

		numRectanglesHorizontal = (int) Math.ceil((width / mBitmapWidth));
		numRectanglesVertical = (int) Math.ceil(height / mBitmapHeight);

		generatePatternBitmap();
	}

	private void generatePatternBitmap(){

		if(getBounds().width() <= 0 || getBounds().height() <= 0) return;
		if (mSrcBitmap == null) return;
		
		Bitmap bitmapCache = mBitmapCache.get(numRectanglesHorizontal * numRectanglesVertical);
		if (bitmapCache != null) {
			mBitmap = bitmapCache;
			return;
		}
		
		Bitmap bitmap = Bitmap.createBitmap(getBounds().width(), getBounds().height(), mSrcBitmap.getConfig());
		Canvas canvas = new Canvas(bitmap);
		
		for (int i = 0; i <= numRectanglesVertical; i++) {
			for (int j = 0; j <= numRectanglesHorizontal; j++) {
				float left = j * mBitmapWidth;
				float top = i * mBitmapHeight;
				canvas.drawBitmap(mSrcBitmap, left, top, mPaint);
			}
		}
		
		mBitmapCache.put(numRectanglesHorizontal * numRectanglesVertical, bitmap);
		mBitmap = bitmap;
	}
}

File: androidcode/src/main/java/com/s16/drawable/PatternDrawable.kt
package com.s16.drawable

import android.annotation.SuppressLint
import android.content.res.Resources
import android.graphics.*
import android.graphics.drawable.Drawable
import androidx.annotation.DrawableRes
import kotlin.math.ceil


class PatternDrawable : Drawable {

    @SuppressLint("UseSparseArrays")
    private val mBitmapCache = HashMap<Int, Bitmap>()

    private val mDrawPaint: Paint = Paint()
    private val mPaint: Paint = Paint()
    private var mBitmapWidth = 0
    private var mBitmapHeight = 0
    private var mSrcBitmap: Bitmap? = null
    private var mBitmap: Bitmap? = null
    private var numRectanglesHorizontal = 0
    private var numRectanglesVertical = 0

    constructor(bitmap: Bitmap) : super() {
        mSrcBitmap = bitmap
        mBitmapWidth = mSrcBitmap!!.width
        mBitmapHeight = mSrcBitmap!!.height
    }

    constructor(res: Resources, @DrawableRes id: Int) : super() {
        if (id > 0) {
            mSrcBitmap = BitmapFactory.decodeResource(res, id)
            mBitmapWidth = mSrcBitmap!!.width
            mBitmapHeight = mSrcBitmap!!.height
        }
    }

    override fun draw(canvas: Canvas) {
        mBitmap?.let {
            canvas.drawBitmap(it, null, bounds, mDrawPaint)
        }
    }

    override fun setAlpha(alpha: Int) {
        throw UnsupportedOperationException("Alpha is not supported by this drawwable.")
    }

    override fun setColorFilter(cf: ColorFilter?) {
        throw UnsupportedOperationException("ColorFilter is not supported by this drawwable.")
    }

    override fun getOpacity(): Int {
        return PixelFormat.OPAQUE
    }

    override fun getIntrinsicWidth(): Int {
        return mBitmapWidth
    }

    override fun getIntrinsicHeight(): Int {
        return mBitmapHeight
    }

    override fun onBoundsChange(bounds: Rect) {
        super.onBoundsChange(bounds)
        val height: Int = bounds.height()
        val width: Int = bounds.width()
        numRectanglesHorizontal = ceil((width / mBitmapWidth).toDouble()).toInt()
        numRectanglesVertical = ceil((height / mBitmapHeight).toDouble()).toInt()
        generatePatternBitmap()
    }

    private fun generatePatternBitmap() {
        if (bounds.width() <= 0 || bounds.height() <= 0) return
        if (mSrcBitmap == null) return
        val bitmapCache = mBitmapCache[numRectanglesHorizontal * numRectanglesVertical]
        if (bitmapCache != null) {
            mBitmap = bitmapCache
            return
        }
        val bitmap = Bitmap.createBitmap(bounds.width(), bounds.height(), mSrcBitmap!!.config)
        val canvas = Canvas(bitmap)
        for (i in 0..numRectanglesVertical) {
            for (j in 0..numRectanglesHorizontal) {
                val left = (j * mBitmapWidth).toFloat()
                val top = (i * mBitmapHeight).toFloat()
                canvas.drawBitmap(mSrcBitmap!!, left, top, mPaint)
            }
        }
        mBitmapCache[numRectanglesHorizontal * numRectanglesVertical] = bitmap
        mBitmap = bitmap
    }
}